### ISSUE TITLE
Add option to not display result of `ddev exec`

### ISSUE THREAD

**Issue by bserem** [drud/ddev#3518]
**Is your feature request related to a problem? Please describe.**
I am working on adding [support for DDEV on GrumPHP (https://github.com/phpro/grumphp/pull/972/files)
At the moment `ddev exec` outputs the command that failed to execute, either in full or in json with `ddev exec -j`.

**Full Output of a failing commit**
![image](https://user-images.githubusercontent.com/520237/149293149-7e3e70cf-eda5-41f6-a178-60ca002aa0c7.png)
**JSON Output of a failing commit**
![image](https://user-images.githubusercontent.com/520237/149293406-73644e94-3ec5-4980-8bc6-729f1edd0c46.png)

**Describe the solution you'd like**
I'd like  a `ddev exec` option to not output anything, GrumPHP will be red or green anyway and that's all the verbosity that is required.
Something like `ddev exec --no-output COMMAND`

**Describe alternatives you've considered**
Run `git` commands from inside the container, but it is not really an option because one would have to copy over ssh keys, git config (easy) and gpg keys (I don't know how to do this yet).

**Additional context**
Grum seems to be widely adopted. If we manage to get https://github.com/phpro/grumphp/pull/972 in a good state it would mean out-of-the-box support for DDEV. PR over there is working at the moment, though I'd 'love to improve it to run both inside and outside of ddev containers.

**Comment by jonaseberle** [comment]
Would `ddev exec COMMAND &> /dev/null` work for you?

**Comment by bserem** [comment]
I tried it, it doesn't work and gives bad results. What happens is that:

1. files get commited regardless of the output of grumphp
2. grumphp is running after the files are commited
3. I get full verbosity (as describe is first post) but I do not return on the terminal until I press enter

Without it things work fine, but I just have that extra verbosity.

**Comment by jonaseberle** [comment]
Can you show your implementation? How do you call `ddev exec`?

**Comment by bserem** [comment]
Look here please: https://github.com/phpro/grumphp/pull/972/files#diff-11109f893cd4e3ac1046f8b102680ee842042e472dcddb9cfd80db2f0b699ff1

**Comment by bserem** [comment]
I did, as I explain in https://github.com/drud/ddev/issues/3518#issuecomment-1011967625
it doesn't work properly.

**Comment by jonaseberle** [comment]
I don't see `&> /dev/null` in your code. 
And please elaborate "does not work" means. We can not help you like that.

**Comment by bserem** [comment]
I tried `&> /dev/null` locally, since it breaks things I did not update my PR.

As I explained in the above comment, **when using `&> /dev/null`:**
1. files get commited **regardless** of the output of grumphp
2. grumphp is running **after** the files are commited
3. I get full verbosity (as describe is first post) but I do not return on the terminal until I press enter

As it is in my PR the output is the one displayed in the screenshots I provided. It works fine, but it is more verbose than it should be.

**Comment by jonaseberle** [comment]
It might be easier if you try these commands without `grumphp` to understand what's going on.
`ddev exec` does not return with the exit code of the command for example (something we could improve), but just SUCCESS (0) or FAILURE (1). This should be enough for your case though.

```
$ ddev exec "git ...."
$ echo $?
$ ddev exec "git .... &> /dev/null"
$ echo $?
```

Also that your .git-hook (pre-commit?) runs after the commit is pretty weird but also something that is not about ddev really.

**Comment by bserem** [comment]
Adding `&> /dev/null` in the command completely hides the grum output.
![image](https://user-images.githubusercontent.com/520237/149336183-e501719f-43af-45ee-ad33-b259008c9721.png)


I really don't feel that this is the way to tackle this. A flag for `ddev exec` makes more sense in this case imho.

**Comment by bserem** [comment]
Yup, the idea is that the project lives completely inside DDEV, and the host machine has only docker+ddev.
I already know how to make the commands work **inside and outside of ddev**, but when I am outside (in the host) I get this extra verbose feedback from `ddev exec`.

**Comment by jonaseberle** [comment]
Maybe the option should be called `--no-error-output` or maybe we should get rid of the additional error output entirely @rfay ?

**Comment by rfay** [comment]
Could you please demonstrate this with a trivial command instead of a fancy setup? 

Is it just stderr you want to be rid of? Or stdout + stderr? 

For example, if I do this command, how would you want it to behave differently?
```
$ ddev exec "ls junker99"
ls: cannot access 'junker99': No such file or directory
Failed to execute command ls junker99: exit status 2
```

The "ls: cannot access 'junker99': No such file or directory" is stderr coming from the "ls" command. 
The "Failed to execute" is coming from ddev. 

Please say exactly how you'd like it to change.

**Comment by bserem** [comment]
Please help me clarify:
- `ls: cannot access 'junker99': No such file or directory` is **stdout**
- `Failed to execute command ls junker99: exit status 2` is **stderr**, right?

In any case, I could use an option to **not show** the `Failed to execute command...`
Does this make sense?

**Comment by jonaseberle** [comment]
It's complicated.

Plain login shell `ls junker99` outputs its error ("ls: cannot access...") to **stderr** and exits with exit code 2 (indicating "serious trouble" as per the `man` page).

`docker-compose exec` turns **stderr** into **stdout** (as far as I know this is by design. `docker exec` does not do that). It exits with code 2.
A little plain docker-compose test:
```
cat <<EOF > docker-compose.yaml
services:
  web: 
    image: debian
    command: sh -c 'while true; do sleep 1; done'
EOF

docker-compose up -d
echo "docker-compose plain:"
docker-compose exec web ls junker99 # outputs "ls: junker99 not found"
echo "docker-compose without stderr:"
docker-compose exec web ls junker99 2> /dev/null # although the output is stderr, it still outputs "ls: junker99 not found"
echo "Exit code: $?" # outputs "Exit code: 2"
# you can do tests with `docker exec` here. Depending on your directory/docker-compose project name
# something like "docker exec tmp_web_1 ls junker99 2> /dev/null"
docker-compose down
```

`ddev exec junker99` outputs the **stderr** of the command to **stdout**, then adds its own output if there has been an error ("Failed to execute <called command>: exit status 2") to **stderr** and exits with code 1.

So long story short: It would be nice to not have ddev output the erroring command line.
I would fancy even turning that around and making it the norm.
In places where this has been helpful debugging information (exec: hooks) it could still be output by the caller (or we could have a `--show-command-on-error`).
Having the correct exit code is not what the issue starter requested and (probably) another story.

**Comment by rfay** [comment]
First @bserem 

> In any case, I could use an option to not show the Failed to execute command...

I wouldn't be opposed to that. Maybe by default. And maybe ddev could actually parse the docker-compose or docker output to determine the error code and return that. It's an ugly hack, but might be worth it.

For now, could you use `ddev exec ... | grep -v "Failed to execute command"` - would that solve your problem?

---

It's really annoying that docker-compose doesn't return the actual return value. 

There are two basic classes of problem here:
1. `docker-compose exec` doesn't return the actual exit code but `docker exec` does.
2. Neither docker-compose nor docker clients properly keep stdout and stderr sorted. They return combined stdout and stderr as stdout. 

The first is fairly easy to sort - we could change `ddev exec` to use the direct-docker-exec technique instead of `docker-compose exec` and I think this would be preferable. The underlying code is already there and used for some other things.

The second is either a problem with the clients or with the docker api, I don't know which. 

There are [several related issues](https://github.com/docker/compose/issues?q=is%3Aissue+is%3Aopen+stderr), most important seems to be https://github.com/docker/compose/issues/8098

**Comment by thaJeztah** [comment]
I think https://github.com/docker/compose/issues/8098 is because `docker-compose exec` defaults to using a tty (so is the equivalent to `docker exec -it ....`

Does it work if it would use `docker-compose exec -T ... ` ?

**Comment by thaJeztah** [comment]
> Neither docker-compose nor docker clients properly keep stdout and stderr sorted. They return combined stdout and stderr as stdout.

Issue related to that is https://github.com/moby/moby/issues/31706. If I recall correctly, the root cause for that is that on Linux [`stdout` is "buffered", but `stderr` not](https://stackoverflow.com/questions/19990589/stderr-and-stdout-not-buffered-vs-buffered). Docker collects data from both, but due to one of them being buffered, output may not appear directly and because of that appear out of order.

**Comment by jonaseberle** [comment]
> Does it work if it would use `docker-compose exec -T ... ` ?

Yes, that keeps **stderr** as **stderr** with my example (tested with docker-compose v1, v2.2.0 and v2.2.3). Thank you for that! I never knew it had something to do with "pseudo-TTY"!

So `docker-compose -T` does everything right (stderr stays stderr, exit code is correct) if I understand the problem correctly.

```
# just something to play with. If stderr stays stderr, it should not get output
docker-compose exec -T web \
  sh -c 'echo "to stdout"; echo "to stderr" 1>&2; exit 2' \
  2> /dev/null
echo "exit code: $?"
# outputs: "to stderr" "exit code 2"
```

**Comment by rfay** [comment]
`ddev exec` already uses `-T` if in terminal context, see https://github.com/drud/ddev/blob/11f2eeec84f8de831d9ee5cb6b4df67ac41ff360/pkg/ddevapp/ddevapp.go#L1240-L1242

Most `ddev exec` are assumed to be interactive these days, so it's important not to mess with that.

OTOH, it wouldn't surprise me if ddev isn't properly processing stdout/stderr, a PR would be welcome. 

It's using dockerutils.ComposeWithStreams() in the normal situation, a quick review doesn't show problems with stdout/stderr.

**Comment by rfay** [comment]
@bserem not sure you saw my question above:

> For now, could you use ddev exec ... | grep -v "Failed to execute command" - would that solve your problem?

**Comment by bserem** [comment]
I've seen it, I'll test it on Monday. Sorry for the delay

**Comment by bserem** [comment]
Sadly, no luck with grepping the output:
![image](https://user-images.githubusercontent.com/520237/149742503-df752060-e74c-475c-8b97-c51fbe636af8.png)

_In the case of the git hook, adding a grep results in a commit, bit this is unrelated to DDEV._

This turns out to be tougher that I anticipated, and it is a niche thing as it seems. I will update the PR on GrumPHP about this discussion here.

**Comment by jonaseberle** [comment]
That should be 
```bash
ddev exec 'ls randomdir' 2>&1 | grep -v 'Failed to execute'
```
(you want to hide what *ddev* outputs to **stderr** so we merge it first into **stdout**)

**Comment by jonaseberle** [comment]
or currently just as good:
```bash
ddev exec 'ls randomdir' 2> /dev/null
```

**Comment by bserem** [comment]
Both of the above work for simple commands (like `ls random`) with simple output.
None of the above work for more the complex commands for the git-hooks in from Grum.

**Comment by rfay** [comment]
Could you update here and in the title what it is you really want to accomplish? You just want `ddev exec` not to add the "Failed to execute command"?  

https://github.com/drud/ddev/blob/665ce0e0042248b273836e08313ff85491dd5415/cmd/ddev/cmd/exec.go#L52-L54

You could start by just removing that and `make` and see if it solves your need.

**Comment by bserem** [comment]
I haven't forgotten about this, I am just buried deep at the moment so not active with this one. I don't feel that confident about compiling ddev right now, as this would tamper with my work-machine.

**Comment by jonaseberle** [comment]
@bserem I felt so, too. But it is actually pretty easy. After checking out the ddev repo or your fork you just run `make`. The new `ddev` binary is then in `.gotmp/bin/*/ddev`.
That is totally isolated and does not affect your normal installation.
For testing, `ddev poweroff` your normal installation and `<path to ddev-git>/.gotmp/bin/*/ddev start` your test project.

**Comment by bserem** [comment]
@jonaseberle nice to know, thanks for the heads up. I tried it, but the web container won't start.


```
$ ~/Development/ddev/.gotmp/bin/linux_amd64/ddev logs -s web
[...]
Instrumentation is opted in, but SegmentKey is not available.
```

**Comment by rfay** [comment]
You don't have the private SegmentKey for complication, and this is just a warning for release-time problems, makes no difference. You can `ddev config global --instrumentation-opt-in=false` if it annoys you too much.

**Comment by bserem** [comment]
Doesn't annoy me at all, I just thought this was the reason ddev didn't start. 

I'll try with a new project first thing in the morning

**Comment by bserem** [comment]
I can't get 2 versions of DDEV to co-exist on my machine without issues, so I still can't test this. Sorry.
At the moment time is also very limited for me to build a new machine just for tests. Maybe someone jumps in and gives some feedback.

Thanks

**Comment by rfay** [comment]
It's very easy to run multiple versions of ddev. It's just a binary, so you can just put it anywhere in your path. You can also rename it. Probably the easiest thing to do is to just link a version that you build into ~/bin and add ~/bin to your path before where ddev is on your system, probably /usr/local/bin.

**Comment by bserem** [comment]
Yeah, I got up to that part, but the second version (custom build to not display the result) does not start:
```
Container ddev-test-dba  Started 
Container ddev-test-web  Started 
Container ddev-test-db  Started 
Container ddev-router  Running 
Failed to start test: container failed to become healthy: err=container 'web' exited, please use 'ddev logs -s web' to find out why it failed 
```

```
$ ~/Development/ddev/.gotmp/bin/linux_amd64/ddev logs -s web
+ set -o errexit nounset pipefail
+ rm -f /tmp/healthy
+ DDEV_PHP_VERSION=7.4
+ DDEV_WEBSERVER_TYPE=nginx-fpm
+ '[' -n 7.4 ']'
+ update-alternatives --set php /usr/bin/php7.4
+ ln -sf /usr/sbin/php-fpm7.4 /usr/sbin/php-fpm
+ export PHP_INI=/etc/php/7.4/fpm/php.ini
+ PHP_INI=/etc/php/7.4/fpm/php.ini
+ '[' '!' -z ']'
+ '[' -d /mnt/ddev_config/php ']'
+ '[' -d /mnt/ddev_config/nginx_full ']'
+ rm -rf /etc/nginx/sites-enabled
+ cp -r /mnt/ddev_config/nginx_full /etc/nginx/sites-enabled/
+ '[' -d /mnt/ddev_config/apache ']'
+ rm -rf /etc/apache2/sites-enabled
+ cp -r /mnt/ddev_config/apache /etc/apache2/sites-enabled
+ '[' php = backdrop ']'
+ '[' php = drupal6 ']'
+ '[' php = drupal7 ']'
++ id -un
++ id -gn
+ printf '\nexport APACHE_RUN_USER=bserem\nexport APACHE_RUN_GROUP=bserem\n'
+ a2enmod access_compat alias auth_basic authn_core authn_file authz_core authz_host authz_user autoindex deflate dir env filter mime mpm_prefork negotiation reqtimeout rewrite setenvif status
Considering dependency authn_core for access_compat:
Module authn_core already enabled
Module access_compat already enabled
Module alias already enabled
Considering dependency authn_core for auth_basic:
Module authn_core already enabled
Module auth_basic already enabled
Module authn_core already enabled
Module authn_file already enabled
Module authz_core already enabled
Considering dependency authz_core for authz_host:
Module authz_core already enabled
Module authz_host already enabled
Considering dependency authz_core for authz_user:
Module authz_core already enabled
Module authz_user already enabled
Module autoindex already enabled
Considering dependency filter for deflate:
Module filter already enabled
Module deflate already enabled
Module dir already enabled
Module env already enabled
Module filter already enabled
Module mime already enabled
Considering conflict mpm_event for mpm_prefork:
Considering conflict mpm_worker for mpm_prefork:
Enabling module mpm_prefork.
Module negotiation already enabled
Module reqtimeout already enabled
Enabling module rewrite.
Module setenvif already enabled
Module status already enabled
To activate the new configuration, you need to run:
  service apache2 restart
+ a2enconf charset localized-error-pages other-vhosts-access-log security serve-cgi-bin
Conf charset already enabled
Conf localized-error-pages already enabled
Conf other-vhosts-access-log already enabled
Conf security already enabled
Conf serve-cgi-bin already enabled
+ '[' nginx-fpm = apache-fpm ']'
+ '[' false = true ']'
+ disable_xdebug
Disabled xdebug
+ phpenmod assert
+ ls /var/www/html
+ sudo mkdir -p /mnt/ddev-global-cache/terminus/cache
+ sudo mkdir -p /mnt/ddev-global-cache/bashhistory/test-web /mnt/ddev-global-cache/mysqlhistory/test-web
++ id -u
++ id -g
+ sudo chown -R 1000:1000 /mnt/ddev-global-cache/
+ '[' -d /mnt/ddev_config/.homeadditions ']'
+ cp -r /mnt/ddev_config/.homeadditions/. /home/bserem/
cp: '/mnt/ddev_config/.homeadditions/./.gitconfig' and '/home/bserem/./.gitconfig' are the same file
Instrumentation is opted in, but SegmentKey is not available. 
```

**Comment by rfay** [comment]
You have something very, very wrong going on there, maybe it's an obscure bug or something. But...
`cp: '/mnt/ddev_config/.homeadditions/./.gitconfig' and '/home/bserem/./.gitconfig' are the same file`

means that perhaps you have used a symlink for your .gitconfig? Or... ??? 

I'd love to know how you created this. It's trying to copy /mnt/ddev_config/.homeadditions into the /home directory and it's finding they're the same file. That means a link of some type has been used.

If you just remove the .gitconfig that you have placed in either ~/.ddev/homeadditions or `<project>/.ddev/homeadditions` it will solve this, but I don't know how you can even get this.

**Comment by bserem** [comment]
Oh! Yes, I did have a symlink:
```
$ ls -la .gitconfig
.gitconfig -> /home/bserem/.gitconfig
```
I removed it for our tests.

---

So, by following your suggestion (https://github.com/drud/ddev/issues/3518#issuecomment-1015468352) I got this to work:
![image](https://user-images.githubusercontent.com/520237/155616081-38509a31-1a97-4eb4-9097-f095b20a64b7.png)

Here is the old result for reference:
![image](https://user-images.githubusercontent.com/520237/155616662-33a8d8ab-2165-4164-a887-ffd6d57d16d4.png)

**Comment by rfay** [comment]
Your existing symlink could never work anyway, because it was trying to reach into the host when it's being used in the container. In the future, you'll want to *copy* your .gitconfig in there.

**Comment by jonaseberle** [comment]
That doesn't help though if you want to use the return value of a program run through `ddev exec`. 

If you want to run a program within a ddev container and use it like you are used from local commands or docker/ssh/..., you would currently have to wrap `ddev exec` to get rid of the error output.

So I still think this (and the stderr > stdout conversion) is not so nice currently.

This is the closest you get to native behaviour IMHO: 
```
docker exec "$(ddev describe -j | jq -r '.raw.services.web.full_name')" ls randomdir
```
=> stderr stays stderr, return value is what the command has set and no additional error output from the "wrapper program" (ddev).

**Comment by rfay** [comment]
@gilbertsoft apparently @bserem wasn't just bothered by the return code - I think we talked about that at length. It was the actual stderr output that bothered him.

**Comment by rfay** [comment]
It seems to me like this either got solved or isn't useful in most workflows, but I'm going to close this for now, happy to continue the conversation here or elsewhere.

**Comment by Coveted** [comment]
The answer made itself known https://preprocessor.space/questions/displaying-search-results-based-on-the-id-value-retrieved-from-a-json-file

**Comment by andreashager** [comment]
@rfay I found this issue and I have a similar problem by now - we have a a command (PHPStan) that returns a JSON String in the command line. If this command does not success (because an error is found in the code) - the exit code of the programm is 1 and the resulting JSON is displayed.

```shell
Failed to execute command `phpstan analyse -c /var/www/html/phpstan.neon --error-format=json --no-interaction --memory-limit=1G`: exit status 1
```

This output prevents the JSON from beeing correct decoded - maybe another option for `ddev exec` to supress this warning would be a nice to have feature 😊

**Comment by andreashager** [comment]
What I would like to have is that i can control those output (https://github.com/ddev/ddev/blob/main/cmd/ddev/cmd/exec.go#L72) with a parameter 😊

So for example `ddev exec -q phpstan analyse -c /var/www/html/phpstan.neon --error-format=json --no-interaction --memory-limit=1G` - then only the output of PHPStan should be available and "Failed to execute command..." should not appear 😊 

> How about using phpstan analyse -c /var/www/html/phpstan.neon --error-format=json --no-interaction --memory-limit=1G || true  ?

If I execute it in command line, the command works without any DDEV output - but we use a third party extension that executes that command and I cannot modify the analyse command via config 😕

**Comment by jonaseberle** [comment]
Removing that "Failed to ...: exit status XX" would still be the right thing to do IMHO. This would make calling stuff with `ddev exec` behave the same way as calling any other command.

**Comment by stasadev** [comment]
I don't know why it wasn't proposed, but here it is, for example:

```bash
$ ddev exec "echo >&2 stderr; foobar"
stderr
bash: line 1: foobar: command not found

Failed to execute command `echo >&2 stderr; foobar`: exit status 127
```

And with `2>/dev/null`:

```bash
$ ddev exec "echo >&2 stderr; foobar" 2>/dev/null
stderr
bash: line 1: foobar: command not found
```

**Comment by andreashager** [comment]
The problem with this is that the command that gets executed (the last part) cannot be modified because its from a third party extension. So I am not able to add `2> /dev/null` to the command 😕

I would like the execute any command without DDEV interfering the output of the command 😊

**Comment by rfay** [comment]
> If I execute it in command line, the command works without any DDEV output - but we use a third party extension that executes that command and I cannot modify the analyse command via config 😕

It's possible that `ddev exec phpstan` isn't what you want to do in the first place. Maybe you want to provide a script in your repo that does what you want and controls output exactly the way you want, or maybe `ddev exec` isn't the right thing in the first place.

Anyway, I think a PR that added a flag to `ddev exec` would be accepted. This isn't very hard stuff.

**Comment by stasadev** [comment]
I added `2>/dev/null` to the `ddev exec`, not to the inner command.

**Comment by jonaseberle** [comment]
@stasadev your example showcases the second problem we have with `ddev exec` (the implicit stderr -> stdout conversion).

**Comment by rfay** [comment]
@jonaseberle is on the job! Looking forward to that PR :)

**Comment by andreashager** [comment]
> [@jonaseberle](https://github.com/jonaseberle) is on the job! Looking forward to that PR :)

@rfay Any progress on this? Should I try a PR instead? 😊

**Comment by jonaseberle** [comment]
No progress from my side.

**Comment by rfay** [comment]
Some thoughts again:

With `bin/do-something-and-fail.sh` in the repository (and executable):
```bash
#!/bin/bash

echo "this is some stdout"
echo "this is some stderr" >/dev/stderr
exit 1
```

```
# Run a script ignoring stdout, stdin, or failure
$ ddev exec bash -c "do-something-and-fail.sh >/dev/null 2>&1 || true"
$ echo $?
0

# Run a script without paying attention to failure but still showing the output
$ ddev exec bash -c "do-something-and-fail.sh || true"
this is some stdout
this is some stderr
$ echo $?

0

# Run a script without showing output at all but still caring about the exit status
$ ddev exec bash -c "do-something-and-fail.sh >/dev/null 2>&1"
Failed to execute command `bash -c "do-something-and-fail.sh >/dev/null 2>&1"`: exit status 1
$ echo $?
1

```

**Comment by stasadev** [comment]
Quiet flag `ddev exec -q command` is implemented in:

- #7385

To install DDEV HEAD: https://ddev.readthedocs.io/en/stable/developers/building-contributing/#testing-latest-commits-on-head
